numba_dppy/utils/array_utils.py

```diff
@@ -0,0 +1,205 @@
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_from_host(obj_memview)
+
+
+def copy_from_usm_backed(usm_backed, obj):
+    """
+    Copy from USM backed data to supported objects.
+
+    Args:
+        usm_backed: Object conformant to __sycl_usm_array_interface__.
+        obj (numpy.ndarray): Numpy ndarray, the data will be copied into.
+
+
+    Raises:
+        TypeError: If any argument is not of permitted type.
+        ValueError: If size of data does not match.
+    """
+    usm_mem = is_usm_backed(usm_backed)
+    if usm_mem is None:
+        raise TypeError("Source is not USM backed.")
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_to_host(obj_memview)
+
+
+def as_usm_backed(obj, queue=None, usm_type="shared", copy=True):
+    """
+    Determine and return a SYCL device accesible object.
+
+    We try to determine if the provided object conforms to
+    __sycl_usm_array_interface__. If not, we create a USM memory of `usm_type`
+    and try to copy the data obj holds. If obj is not already backed by USM,
+    only numpy.ndarray is supported currently.
+
+    Args:
+        obj: Object to be tested and data copied from.
+        usm_type: USM type used in case obj is not already backed by USM.
+        queue (dpctl.SyclQueue): SYCL queue to be used in case obj is not
+            already USM backed.
+        copy (bool): Flag to determine if we copy data from obj.
+
+    Returns:
+        obj: USM backed memory.
+
+    Raises:
+        TypeError:
+            1. If obj is not already backed by USM and is not of type
```

Comment:
We want either an object that already was allocated on USM memory or an numpy ndarray. 

That is the intent of what we check. Right?

The object does not have to be a numpy ndarray backed by USM.